man/man1/lgt2txt.1
.TH lgt2txt 1 "September 16, 2019" "Logtalk 3.30.0" "Logtalk Documentation"

.SH NAME
lgt2txt \- Logtalk script for converting XML documenting files into text files

.SH SYNOPSIS
.B lgt2txt
[\fB-d \fIdirectory\fR]
[\fB-p \fIprocessor\fR]
.br
.B lgt2txt
.B \-h

.SH DESCRIPTION
\fBlgt2txt\fR is a Logtalk script for converting all Logtalk XML documenting files in the current directory into text files. Assumes UTF-8 encoding.

.SH OPTIONS
.TP
.B \-d
.I directory
Output directory for the text files.
.TP
.B \-p
.I processor
XSLT processor.
.TP
.B \-h
Prints a summary of the script options.

.SH REQUIREMENTS
Requires a compatible XSLT processor. See the script help screen for more information.

.SH ENVIRONMENT
.TP
.B LOGTALKHOME
Specifies the location of the Logtalk installation.
.TP
.B LOGTALKUSER
Specifies the location of the Logtalk user directory.

.SH FILES
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_entity.dtd
The Logtalk entity DTD file.
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_index.dtd
The Logtalk index DTD file.
.TP
.BI $LOGTALKUSER/tools/lgtdoc/xml/custom.ent
The Logtalk custom entities file.
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_entity.xsd
The Logtalk entity XML Schema file.
.TP
.BI $LOGTALKHOME/tools/lgtdoc/xml/logtalk_index.xsd
The Logtalk index XML Schema file.
.TP
.BI $LOGTALKUSER/tools/lgtdoc/xml/logtalk_entity_to_txt.xsl
The Logtalk entity XSLT file.
.TP
.BI $LOGTALKUSER/tools/lgtdoc/xml/logtalk_index_to_txt.xsl
The Logtalk index XSLT file.

.SH "SEE ALSO"
The Logtalk website at https://logtalk.org/
.PP
The Logtalk User and Reference Manuals at \fB$LOGTALKHOME/manuals/index.html\fR
.PP
logtalk_user_setup(1),\ lgt2html(1),\ lgt2xml(1),\ lgt2pdf(1),\ lgt2md(1)

.SH AUTHOR
Paulo Moura <pmoura@logtalk.org>

.SH COPYRIGHT
Copyright (c) 1998-2021 Paulo Moura.


RELEASE_NOTES.md
________________________________________________________________________

This file is part of Logtalk <https://logtalk.org/>  
Copyright 1998-2021 Paulo Moura <pmoura@logtalk.org>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
________________________________________________________________________


RELEASE NOTES
=============


3.44.0 - January ??, 2021
=========================

Documentation
-------------

* IMPROVED: Documentation of the `logtalk::loaded_file_property/2` predicate.

Library
-------

* ADDED: Implementation for the `os::directory_files/2` predicate for Trealla
ProLog.

* FIXED: Implementation for the `os::delete_directory/1` predicate for Trealla
ProLog.

Installers and installation scripts
-----------------------------------

* CHANGED: Renamed the `scripts/macosx` directory to `scripts/macos`.

* ADDED: Windows installer experimental support for creating a shortcut for
Tau Prolog.

* ADDED: A macOS command-file for Trealla ProLog.

* ADDED: A `update_man_html_versions.sh` shell script for updating the HTML
versions of the man pages.

* UPDATED: The macOS installer to also set system-wide default values for the
`LOGTALKHOME` and `LOGTALKUSER` environment variables for the `zsh` shell.


3.43.0 - December 22, 2020
==========================

Logtalk compiler and runtime
----------------------------

* ADDED: Lint check for redundant use of the `call/1` control construct as
argument to the `\+/1` built-in predicate.

* ADDED: Also check correctness of `variable_names/1` options in calls to the
`write_term/2-3` Prolog standard predicates.

* IMPROVED: Compiler performance for protocols.

* IMPROVED: More efficient compiler checking for parameter variables.

* FIXED: Using a overriding meta-predicate template for predicates defined in
`user` listed in `uses/2` directives.

* FIXED: Bug in the `logtalk_make/0-1` predicates when reloading a main file
due to an updated included file when the `clean` flag is turned off and the
`reload` file is set tp `changed` where the main file would not be recompiled
to include the new contents of the included file.

* FIXED: Bug in the `asserta/1` built-in method when the argument is only know
at runtime.

Prolog adapter and integration files
------------------------------------

* ADDED: Adapter file for Trealla Prolog. Experimental. With the help of
Andrew Davison.

* IMPROVED: Adapter files definition of the `{}/1` top-level interpreter
shortcut for loading files to use a single call to the `logtalk_load/1`
built-in predicate.

* UPDATED: The LVM adapter file to take advantage of indexing support for the
internal runtime tables and of the new `atom_hash/2` built-in predicate.

* UPDATED: The LVM integration script to select the standard-top-level.

* UPDATED: The Tau Prolog and Trealla ProLog adapter files for Unicode support.

Documentation
-------------

* ADDED: Man page for the Trealla ProLog integration script, `tplgt`.

* FIXED: Mention of settings file possible locations in the integration script
man pages.

Library
-------

* ADDED: Support for Trealla ProLog to the `os` and `random` libraries.

* ADDED: Library `git` with predicates for accessing a git project current
branch and latest commit data.

* ADDED: Definition for the `os::wall_time/1` predicate for Tau Prolog.

* ADDED: New `jump_all_block/3`, `append/3`, and `map/2-3` predicates to the
`queues` library.

* ADDED: Tests for the `os` library `wall_time/1` and `shell/1-2` predicates.

* UPDATED: The `hook_objects` library `object_wrapper_hook` object to set the
`context_switching_calls` flag to `allow` for the generated objects, enabling
calling the wrapped predicates using the `<</2` context-switching control
construct.

* FIXED: The `os::directory_files/2` predicate for LVM to also return the `.`
and `..` directories as per the `osp` protocol.

* FIXED: Test failure for the `hook_objects` library due to missing setup step.

* FIXED: Fix the `os::shell/1` library predicate for ECLiPSe (broken due to a
bug in its `system/1` built-in predicate).

* FIXED: The `os::command_line_arguments/1` predicate for Tau Prolog.

* FIXED: The `os` library predicates `cpu_time/1` and `wall_time/1` for LVM.

Tools
-----

* CHANGED: The `lgtunit` tool now requires and checks that test identifiers
are ground callable terms.

* ADDED: Predicates `set_text_output/3` and `set_binary_output/3` to the
`lgtunit` tool.

* ADDED: Support for Trealla ProLog to the `debugger`, `lgtunit`, and
`ports_profiler` tools.

* ADDED: Additional test for the `debugger::trace/0` predicate.

* IMPROVED: Advice given by the `tutor` tool for warnings generated by compile
time references to unknown entities.

* IMPROVED: Documentation of the `ports_profiler` tool.

* UPDATED: The `logtalk_tester` script support for the LVM backend.

* FIXED: The `lgtunit::run_test_sets/1` predicate to type-check the test set
objects.

* FIXED: The `lgtunit` tool expansion of `discontiguous/1` directives to apply
only to entity directives.

* FIXED: The `debugger` tool to call the `reset/0` predicate at loading time
to ensure that code compiled in debug mode will also run even when not tracing
or debugging. Thanks to Camilo Correa Restrepo for the bug report.

* FIXED: The `lgtdoc` tool to abolish the temporary objects created to find
inherited predicates when documenting entities.

Examples
--------

* ADDED: Crude solution to the `bench` example to run all benchmarks.

* ADDED: Tests for some of the individual benchmarks in the `bench` example.


tools/lgtdoc/xml/lgt2txt.js
/////////////////////////////////////////////////////////////////////////////
//
//   XML documenting files to plain text conversion script 
//   Last updated on July 5, 2016
//
//   This file is part of Logtalk <https://logtalk.org/>  
//   Copyright 1998-2021 Paulo Moura <pmoura@logtalk.org>
//   
//   Licensed under the Apache License, Version 2.0 (the "License");
//   you may not use this file except in compliance with the License.
//   You may obtain a copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
//   Unless required by applicable law or agreed to in writing, software
//   distributed under the License is distributed on an "AS IS" BASIS,
//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//   See the License for the specific language governing permissions and
//   limitations under the License.
//
/////////////////////////////////////////////////////////////////////////////


if (ScriptEngineMajorVersion() < 5 || ScriptEngineMajorVersion() == 5 && ScriptEngineMinorVersion() < 6) {
	WScript.Echo('Error! WSH 5.6 or later version needed for running this script.');
	WScript.Quit(1);
}

var WshShell = new ActiveXObject("WScript.Shell");

var directory = WshShell.CurrentDirectory;

var processor = "msxsl";
// var processor = "xsltproc";
// var processor = "xalan";
// var processor = "sabcmd";

if (WScript.Arguments.Unnamed.Length > 0) {
	usage_help();
	WScript.Quit(0);
}

var WshSystemEnv = WshShell.Environment("SYSTEM");
var WshUserEnv = WshShell.Environment("USER");

var logtalk_home;
var logtalk_user;

if (WshSystemEnv.Item("LOGTALKHOME"))
	logtalk_home = WshSystemEnv.Item("LOGTALKHOME");
else if (WshUserEnv.Item("LOGTALKHOME"))
	logtalk_home = WshUserEnv.Item("LOGTALKHOME")
else {
	WScript.Echo("Error! The environment variable LOGTALKHOME must be defined first!");
	usage_help();
	WScript.Quit(1);
}

if (WshSystemEnv.Item("LOGTALKUSER"))
	logtalk_user = WshSystemEnv.Item("LOGTALKUSER");
else if (WshUserEnv.Item("LOGTALKUSER"))
	logtalk_user = WshUserEnv.Item("LOGTALKUSER")
else {
	WScript.Echo("Error! The environment variable LOGTALKUSER must be defined first!");
	usage_help();
	WScript.Quit(1);
}

var entity_xslt = logtalk_user + "\\tools\\lgtdoc\\xml\\logtalk_entity_to_txt.xsl";
var index_xslt = logtalk_user + "\\tools\\lgtdoc\\xml\\logtalk_index_to_txt.xsl";

var d_arg = "";
var p_arg = "";

if (WScript.Arguments.Named.Exists("d"))
	d_arg = WScript.Arguments.Named.Item("d");

if (WScript.Arguments.Named.Exists("p"))
	p_arg = WScript.Arguments.Named.Item("p");

var FSObject = new ActiveXObject("Scripting.FileSystemObject");

if (d_arg != "" && !FSObject.FolderExists(d_arg)) {
	WScript.Echo("Error! directory does not exists: " + d_arg);
	WScript.Echo("");
	usage_help();
} else if (d_arg != "")
	directory = d_arg;

if (p_arg != "" && p_arg != "msxsl" && p_arg != "xsltproc" && p_arg != "xalan" && p_arg != "sabcmd") {
	WScript.Echo("Error! Unsupported XSLT processor:" + p_arg);
	WScript.Echo("");
	usage_help();
} else if (p_arg != "")
	processor = p_arg;

if (!FSObject.FileExists(WshShell.CurrentDirectory + "\\logtalk_entity.dtd")) {
	FSObject.CopyFile(logtalk_home + "\\tools\\lgtdoc\\xml\\logtalk_entity.dtd", WshShell.CurrentDirectory + "\\logtalk_entity.dtd");
}

if (!FSObject.FileExists(WshShell.CurrentDirectory + "\\logtalk_index.dtd")) {
	FSObject.CopyFile(logtalk_home + "\\tools\\lgtdoc\\xml\\logtalk_index.dtd", WshShell.CurrentDirectory + "\\logtalk_index.dtd");
}

if (!FSObject.FileExists(WshShell.CurrentDirectory + "\\custom.ent")) {
	FSObject.CopyFile(logtalk_home + "\\tools\\lgtdoc\\xml\\custom.ent", WshShell.CurrentDirectory + "\\custom.ent");
}

if (!FSObject.FileExists(WshShell.CurrentDirectory + "\\logtalk_entity.xsd")) {
	FSObject.CopyFile(logtalk_home + "\\tools\\lgtdoc\\xml\\logtalk_entity.xsd", WshShell.CurrentDirectory + "\\logtalk_entity.xsd");
}

if (!FSObject.FileExists(WshShell.CurrentDirectory + "\\logtalk_index.xsd")) {
	FSObject.CopyFile(logtalk_home + "\\tools\\lgtdoc\\xml\\logtalk_index.xsd", WshShell.CurrentDirectory + "\\logtalk_index.xsd");
}

if (!FSObject.FileExists(directory + "\\logtalk.css")) {
	FSObject.CopyFile(logtalk_user + "\\tools\\lgtdoc\\xml\\logtalk.css", directory + "\\logtalk.css");
}

WScript.Echo("");
WScript.Echo("converting XML files...");

var files = new Enumerator(FSObject.GetFolder(WshShell.CurrentDirectory).Files);

for (files.moveFirst(); !files.atEnd(); files.moveNext()) {
	var file = files.item().name;
	var xslt;
	if (FSObject.GetExtensionName(file) == "xml") {
		WScript.Echo("  converting " + file);
		var txt_file = directory + "\\" + FSObject.GetBaseName(file) + ".txt";
		if (file == "library_index.xml" || file == "directory_index.xml" || file == "entity_index.xml" || file == "predicate_index.xml") {
			xslt = index_xslt;
		} else {
			xslt = entity_xslt;
		}
		switch (processor) {
			case "msxsl" :
				WshShell.Run("msxsl -o \"" + txt_file + "\" \"" + file + "\" \"" + xslt + "\"", true);
				break;
			case "xsltproc" :
				WshShell.Run("xsltproc -o \"" + txt_file + "\" \"" + xslt + "\" \"" + file + "\"", true);
				break;
			case "xalan" :
				WshShell.Run("xalan -o \"" + txt_file + "\" \"" + file + "\" \"" + xslt + "\"", true);
				break;
			case "sabcmd" :
				WshShell.Run("sabcmd \"" + xslt + "\" \"" + file + "\" \"" + txt_file + "\"", true);
				break;
		}
	}
}

WScript.Echo("conversion done");
WScript.Echo("");

WScript.Quit(0);

function usage_help() {
	WScript.Echo("");
	WScript.Echo("This script converts all Logtalk XML files documenting files in the");
	WScript.Echo("current directory to text files");
	WScript.Echo("");
	WScript.Echo("Usage:");
	WScript.Echo("  " + WScript.ScriptName + " [/d:directory] [/p:processor]");
	WScript.Echo("  " + WScript.ScriptName + " help");
	WScript.Echo("");
	WScript.Echo("Optional arguments:");
	WScript.Echo("  d - output directory for the generated files (default is " + directory + ")");
	WScript.Echo("  p - XSLT processor (msxsl, xsltproc, xalan, or sabcmd; default is " + processor + ")");
	WScript.Echo("");
	WScript.Quit(1);
}
